File: c/sorting/quick_sort.c
#include <stdio.h>
#include "../benchmark.c"

/*
* Given references to x and y, swaps them in-place.
*/
static inline void swap(int* a, int* b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}


/*
* Partitions the array between a pivot (array[high]) such
* such that values higher than the pivot are on the right of it.
* and values lower are on the left.
*
* Returns index representing pivot.
*/
int partition(int* array, int low, int high) {
    int pivot = array[high];
    int i = low-1;

    for (size_t j = low; j < high; j++) {
        // Check if we're less than pivot:
        // Swap left and right.
        if (array[j] <= pivot) {
            i++;
            swap(&array[i], &array[j]);
        }
    }

    // Then position the pivot.
    swap(&array[i+1], &array[high]);
    return i+1;
}

/*
* Recursively splits the array until array becomes too small to partition.
*/
void sort(int* array, int low, int high) {
    if (low >= high) {
        return;
    }
    int pivot = partition(array, low, high);
    sort(array, low, pivot-1);
    sort(array, pivot+1, high);
}

void start_quick_sort(int* array, int array_size) {
    benchmark();
    sort(array, 0, array_size-1);
    end_benchmark();
}

File: rust/sorting/quick_sort.rs
// rand = "0.9"

use rand::prelude::*;
use std::env;

// for the use of cargo-single we have to traverse backwards.
include!{"../../random_value.rs"}
include!{"../../../benchmark.rs"}
include!{"../../../util.rs"}

fn partition(array: &mut Vec<u32>, low: usize, high: usize) -> usize {
    let pivot = high;
    let mut i = low;

    for j in low..pivot-1 {
        if array[j] <= array[pivot] {
            array.swap(j, i);
            i+=1;
        }
    }
    array.swap(i, pivot);
    i
}

fn sort(array: &mut Vec<u32>, low: usize, high: usize) {
    if low >= high {
        return;
    }
    let pivot = partition(array, low, high);
    sort(array, low, pivot-1);
    sort(array, pivot+1, high);
}


fn main() {
    let mut args : Vec<String> = env::args().collect();
    warn_arguments(&mut args);

    let array_size : usize = args[1].parse::<usize>().unwrap();
    let mut array = randomize_array(1, args[2].parse::<u32>().unwrap(), array_size);

    benchmark(args[3].parse::<usize>().unwrap(), || sort(&mut array, 1, array_size-1));
}
